File: solver.py
#!/usr/bin/env python2.7

from collections import Counter, defaultdict
from copy import deepcopy
from itertools import combinations as combn
import sys

def row_cells(rc, sudoku):
    """Given coordinates rc, return all the other cells in the same row"""
    r, c = rc
    return [sudoku[(r, i)] for i in range(0, 9) if i != c]

def col_cells(rc, sudoku):
    """Given coordinates rc, return all the other cells in the same column"""
    r, c = rc
    return [sudoku[(i, c)] for i in range(0, 9) if i != r]

def sqr_cells(rc, sudoku):
    """Given coordinates rc, return all the other cells in the same square"""
    r, c = rc
    sc = c / 3
    sr = r / 3
    ret = []
    for i in range(3):
        for j in range(3):
            if (i+sr*3, j+sc*3) == rc:
                continue
            ret.append(sudoku[(i+sr*3, j+sc*3)])
    return ret

def printsudoku(sudoku):
    sys.stdout.write('\n||=====================================||\n')
    for i in range(9):
        sys.stdout.write('||')
        for j in range(9):
            c = sudoku[(i, j)]
            if len(c) > 1:
                p = '?'
            else:
                p = str(list(c)[0])
            sys.stdout.write(' ' + p + ' |')
            if j % 3 == 2:
                sys.stdout.write('|')
        sys.stdout.write('\n')
        if i % 3 == 2:
            sys.stdout.write('||=====================================||\n')
    sys.stdout.write('\n')

def printsudoku_full(sudoku):
    for i in range(9):
        for j in range(9):
            if j > 0:
                sys.stdout.write('\t')
            sys.stdout.write(','.join(str(x) for x in sudoku[(i, j)]))
            if len(sudoku[(i, j)]) == 0:
                sys.stdout.write('??')
        sys.stdout.write('\n')

def n_unsolved(sudoku):
    """Get the number of cells left unsolved"""
    return sum(int(len(c) != 1) for c in sudoku.itervalues())

def read_sudoku_file(fname):
    b = {}
    with open(infile, 'rb') as ihandle:
        for i, line in enumerate(ihandle):
            row = line.split('\t')
            for j, col in enumerate(row):
                col = col.strip()
                if col == '':
                    b[(i, j)] = set(range(1, 10))
                else:
                    b[(i, j)] = set([int(col)])
    return b

def simple_elimination(cell, groups):
    """Use simple elimination methods to solve a cell. Modifies
    `cell'. """
    fixed = set()
    ufixed = []
    for group in groups:
        uf = set()
        for cc in group:
            if len(cc) == 1:
                fixed.update(cc)
            else:
                uf.update(cc)
        ufixed.append(uf)
    if len(cell - fixed) == 0:
        raise Exception('All numbers eliminated!')
    # Eliminate the numbers that this cell cannot be b/c
    # other cells are already set to those numbers
    cell -= fixed
    # Figure out if this is the only cell in a particular row,
    # column, or square that can be a certain number
    for uf in ufixed:
        if len(cell - uf) == 1:
            cell -= uf

def eliminate_with_matching_tuples(cell, groups, max_size=5):
    """Figure out if this cell is part of a pair (or triple, quadruple,
    etc.) that are the only cells containing a certain set of 2 (or 3, 4, etc.)
    items in the row, column, or square. If so, constrain the possibilities
    for this cell to just that set of items.

    Example: this cell is (1,3,4,7) and only one other cell in the row
    contains 3 and 7. Then make this cell just (3,7)."""
    for cells in groups:
        single_overlaps = Counter()
        group_overlaps = defaultdict(int)
        for cc in cells:
            if len(cc) == 1:
                continue
            single_overlaps.update(cell.intersection(cc))
            group_overlaps[tuple(cell.intersection(cc))] += 1
        for tuple_len in range(max_size, 1 , -1):
            tupls = []
            for o, ocount in group_overlaps.iteritems():
                if len(o) == tuple_len and ocount == tuple_len - 1:
                    match = True
                    for tuple_member in o:
                        if single_overlaps[tuple_member] != tuple_len - 1:
                            match = False
                    if match:
                        tupls.append(o)
            if len(tupls) == 1:
                cell.clear()
                cell.update(tupls[0])

def eliminate_with_other_tuples(cell, groups_with_cell, max_size=5):
    """Figure out if there are two other cells in this row,
    column or square that can each only be the same two
    items (e.g. two cells that are both (1,5) ).
    If so, remove those numbers from this cell's list.
    Extended to groups larger than two."""
    for cells in groups_with_cell:
        for size in range(max_size, 1, -1):
            for cell_comb in combn(filter(lambda x: len(x) == size, cells), size):
                tupls = map(tuple, cell_comb)
                match = reduce(lambda x, y: x == y, tupls)
                if match:
                    cell -= cell_comb[0]
            if len(cell) == 1:
                return

def logical_solver(sudoku):
    """Solve the sudoku using logical methods"""
    iteration = 0
    progress = True
    while(progress):
        iteration += 1
        progress = False
        for rc, c in sudoku.iteritems():
            if len(c) == 1:
                continue
            l = len(c)
            groups_with_cell = [row_cells(rc, sudoku), col_cells(rc, sudoku),
                sqr_cells(rc, sudoku)]

            simple_elimination(c, groups_with_cell)
            if len(c) < l:
                progress = True
            if len(c) == 1:
                continue

            eliminate_with_matching_tuples(c, groups_with_cell)
            if len(c) < l:
                progress = True
            if len(c) == 1:
                continue
                
            eliminate_with_other_tuples(c, groups_with_cell)
            if len(c) < l:
                progress = True
            if len(c) == 1:
                continue

    return sudoku, iteration, n_unsolved(sudoku) == 0

def guess_solver(sudoku, prev_chosen=[]):
    """Take a guess at one undetermined cell, then try to solve"""
    cell_to_guess = None
    for n_choices in range(2, 10):
        if cell_to_guess is not None:
            break
        for rc, cell in sudoku.iteritems():
            if len(cell) == n_choices:
                if rc in prev_chosen:
                    continue
                cell_to_guess = rc
                break
    choices = list(sudoku[cell_to_guess])
    for choice in choices:
        sudoku_copy = deepcopy(sudoku)
        sudoku_copy[cell_to_guess] = set([choice])
        solved = False
        try:
            sudoku_copy, niter, solved = logical_solver(sudoku_copy)
        except:
            continue # Failed with this guess, try next choice
        # Probably not the most elegant way to code this, but it has
        # to be able to detect when no solution is found.
        if solved:
            return True, sudoku_copy
        else:
            solved, sudoku_copy = guess_solver(sudoku_copy, prev_chosen + [cell_to_guess])
        if solved:
            return True, sudoku_copy
        else:
            continue
    return False, sudoku_copy # if guessing fails - probably not all choices tried or logic wrong


for infile in sys.argv[1:]:
    start_sudoku = read_sudoku_file(infile)
    print '\nStart: %s (%i)' % (infile, n_unsolved(start_sudoku))
    printsudoku(start_sudoku)

    solved = False
    try:
        end_sudoku, niters, solved = logical_solver(deepcopy(start_sudoku))
    except:
        solved = False
        end_sudoku = start_sudoku

    if not solved:
        end_sudoku = guess_solver(end_sudoku)[1]
        print 'End: (%i)' % n_unsolved(end_sudoku)
        printsudoku(end_sudoku)
    else:
        print 'End: (%i)' % n_unsolved(end_sudoku)
        printsudoku(end_sudoku)

File: solver.jl
#!/usr/bin/env julia

function print_board(b)
    for i in 1:9
        if (i-1) % 3 == 0
            print("||", "="^37, "||\n")
        end
        for j in 1:9
            if (j-1) % 3 == 0
                print("|")
            end
            if length(b[i, j]) == 1
                print("| ", collect(b[i, j])[1], " ")
            else
                print("| ? ")
            end
        end
        print("||\n")
    end
    print("||", "="^37, "||\n\n")
end

function n_unsolved(b)
    n_unsolved = 0
    for cell in b
        if length(cell) != 1
            n_unsolved += 1
        end
    end
    n_unsolved
end

function read_board(fname)
    board = Array{Set{Int},2}(undef, 9, 9)
    open(fname, "r") do handle
        for (i, line) in enumerate(eachline(handle))
            for (j, cell) in enumerate(split(line, "\t"))
                if cell == ""
                    board[i, j] = Set{Int}(1:9)
                else
                    board[i, j] = Set{Int}(parse(Int, cell))
                end
            end
        end
    end
    board
end

function get_rc(b, r, c, row=true)
    ret = Array{Set{Int},1}()
    for i in 1:9
        if (row && i != c) || (!row && i != r)
            if row
                push!(ret, b[r, i])
            else
                push!(ret, b[i, c])
            end
        end
    end
    ret
end

function get_sqr(b, r, c)
    ret = Array{Set{Int},1}()
    sqr_row = div((r-1), 3) * 3 + 1
    sqr_col = div((c-1), 3) * 3 + 1
    for i in sqr_row:(sqr_row + 2)
        for j in sqr_col:(sqr_col+2)
            if !(i == r && j == c)
                push!(ret, b[i, j])
            end
        end
    end
    ret
end


function get_neighbors(b, r, c)
    [get_rc(b, r, c, true), get_rc(b, r, c, false), get_sqr(b, r, c)]
end

function simple_solver!(b)
    n_u = n_unsolved(b)
    while n_u > 0
        for (i, j) in Base.Iterators.product(1:9, 1:9)
            cell = b[i, j]
            if length(cell) == 1
                continue
            end
            neighbors = get_neighbors(b, i, j)
            for n in neighbors
                for c in n
                    if length(c) == 1
                        cc = setdiff(cell, c)
                        if length(cc) < 1
                            throw("too few at $i, $j")
                        end
                        b[i, j] = cc
                        cell = cc
                    end
                end
            end
            if length(cell) == 1 continue end
            r = setdiff(cell, reduce(union, neighbors[1]))
            if length(r) == 1
                b[i, j] = Set{Int}(r)
                continue
            end
            c = setdiff(cell, reduce(union, neighbors[2]))
            if length(c) == 1
                b[i, j] = Set{Int}(c)
                continue
            end
            s = setdiff(cell, reduce(union, neighbors[3]))
            if length(s) == 1
                b[i, j] = Set{Int}(s)
                continue
            end
        end
        n_u_ = n_unsolved(b)
        if n_u_ == n_u
            # No progress
            break
        end
        n_u = n_u_
    end
end

function solve(b, guess_cells=[])
    guess_cell = (0, 0)
    for n_to_guess in 2:9
        for (i, j) in Base.Iterators.product(1:9, 1:9)
            if length(b[i, j]) != n_to_guess || (i, j) in guess_cells
                continue
            else
                guess_cell = (i, j)
                break
            end
        end
    end
    b_cpy = deepcopy(b)
    for c in b[guess_cell[1], guess_cell[2]]
        b_cpy = deepcopy(b)
        b_cpy[guess_cell[1], guess_cell[2]] = Set{Int}(c)
        try
            simple_solver!(b_cpy)
        catch
            continue
        end
        if n_unsolved(b_cpy) == 0
            return b_cpy
        else
            solution = solve(b_cpy, push!(copy(guess_cells), guess_cell))
            if n_unsolved(solution) == 0
                return solution
            else
                continue
            end
        end
    end
    ## If a guess at a higher level was wrong
    return b_cpy
end

function main()
    for fname in ARGS
        board = read_board(fname)
        print("\n", "Start: ", fname, " (", n_unsolved(board), ")\n\n")
        print_board(board)
        solution = solve(board)
        print("\n", "End: ", fname, " (", n_unsolved(solution), ")\n\n")
        print_board(solution)
    end
end

main()
